File: Linked_Lists/Doubly/doublyLL.py
"""
Author: Bryce Calhoun
Description: Doubly Linked List implementation with operations: insert(at the end), remove, search, display forward,
             display backward.

"""


class Node:
    """
    Description: A class defintion for a single node in the list.
                 Comes with previous and next fields for nodes and
                 a value field for the value of the given node.

    """
    def __init__(self, value=None, previous=None, nxt=None):
        self.value = value
        self.previous = previous
        self.nxt = nxt

    # ACCESSOR
    def get_value(self):
        return self.value

    # MUTATOR
    def set_value(self, value):
        self.value = value


class doublyLL:
    """
    Description: Class defintion for the doubly linked list.
                 A head member to track the front of the list,
                 as well as a size member to track to number 
                 of elements.

    """
    def __init__(self, head=None, size=0):
        self.head = head
        self.size = size

    # ACCESSOR
    def get_size(self):
        return self.size

    def insert(self, value):
        """
        Description: Function to insert a node into the doubly-linked
                     list given the value to insert.
                     TIME: O(n).

        """
        node = Node()
        node.set_value(value)
        if self.head is None:
            node.previous = None
            node.nxt = None
            self.head = node
        else:
            current = self.head
            while current.nxt is not None:
                current = current.nxt
            current.nxt = node
            node.previous = current
        self.size += 1
        return

    def search(self, value):
        """
        Description: Function to search for a node in the list given
                     the value to search for. The function will return the node.
                     TIME: O(n).

        """
        if self.head is None:
            print("empty list")
            return None
        else:
            current = self.head
            while current is not None:
                if current.get_value() == value:
                    return current
                current = current.nxt
            print("\nvalue not found\n")
            return None

    def remove(self, value):
        """
        Description: Function to remove a node from the list, given the value of
                     the node. Uses the search() function as a helper.
                     TIME: O(n).

        """
        node = self.search(value)
        if node is None:
            print("\nremoval error\n")
        else:
            if node.nxt is None:
                node.previous.nxt = None
                node.previous = None
            elif node.nxt is not None and node.previous is not None:
                node.previous.nxt = node.nxt
                node.nxt.previous = node.previous
                node.previous = None
                node.nxt = None
            else:
                self.head = node.nxt
                node.nxt.previous = None
                node.nxt = None
            print("value removed: ", value)

        return

    def display_forward(self):
        """
        Description: Function to display the contents of an instance of the
                     doubly-linked list in a head-first order.
                     TIME: O(n).

        """
        if self.head is None:
            print("\nnothing to see\n")
            return
        else:
            current = self.head
            while current is not None:
                print("\n", current.get_value(), "\n")
                current = current.nxt
            return

    def display_backward(self):
        """
        Description: Function to display the contents of an instance of the
                     doubly-linked list in an end-first order.
                     TIME: O(2n)/O(n).

        """
        if self.head is None:
            print("\nnothing to see\n")
            return
        else:
            current = self.head
            while current.nxt is not None:
                current = current.nxt
            while current is not None:
                print("\n", current.get_value(), "\n")
                current = current.previous
            return

File: Linked_List/Doubly/DoublyLL.java
/*
** Author: Bryce Calhoun
** description: Implementation of a doubly linked list of node structures.
*/

public class DoublyLL{
/*
Description: Class definition for the doubly linked list which uses both head a tail sentinels
*/

    private int size;
    private Node head_sentinel;
    private Node tail_sentinel;
    
    public DoublyLL(){
    /*
    Description:  default constructor for the doubly linked list object
    */
        this.head_sentinel = new Node(null, null, 0);
        this.tail_sentinel = new Node(null, null, 0);
        this.size = 0;
    }
    public class doublyLLException extends Exception{
        public doublyLLException(String error){
            super(error);
        }
    }
    // Accessor for the size member variable
    public int get_size(){
        return this.size;
    }
    public int is_empty(){
        if(this.size == 0){
            return 1;
        }
        return 0;
    }
    public void insert_front(int value){
    /*
    Description: Function to insert a node into the front of the doubly linked list
    Input(s): Integer value
    Output(s): None
    */
        if(this.is_empty() == 1){
            Node newNode = new Node(this.head_sentinel, this.tail_sentinel, value);
            this.head_sentinel.next = newNode;
            this.tail_sentinel.previous = newNode;
        }
        else{
            Node newNode = new Node(this.head_sentinel, this.head_sentinel.next, value);
            this.head_sentinel.next = newNode;
            newNode.next.previous = newNode;
        }
        this.size++;
    }
    public int remove_front(){
    /*
    Description: Function to remove a node from the front of the doubly linked list and return its value
    Input(s): None
    Output(s): Integer value
    */
        if(this.is_empty() == 1){
            System.out.println("\nEMPTY LIST\n");
            return 0;
        }
        else{
            int value = this.head_sentinel.next.get_value();
            this.head_sentinel.next = this.head_sentinel.next.next;
            this.head_sentinel.next.previous = this.head_sentinel;

            this.size--;
            return value;
        }
        
    }
    public void insert_back(int value){
    /*
    Description: Function to insert a node into the back of the doubly linked list.
    Input(s): Integer value
    Output(s): None
    */
        if(this.is_empty() == 1){
            Node newNode = new Node(this.head_sentinel, this.tail_sentinel, value);
            this.head_sentinel.next = newNode;
            this.tail_sentinel.previous = newNode;
        }
        else{
            Node newNode = new Node(this.tail_sentinel.previous, this.tail_sentinel, value);
            this.tail_sentinel.previous.next = newNode;
            this.tail_sentinel.previous = newNode;
        }
        this.size++;
    }
    public int remove_back(){
    /*
    Description: Function to remove a node from the back of the doubly linked list and return its value
    Input(s): None
    Output(s): Integer value
    */
        if(this.is_empty() == 1){
            System.out.println("\nEMPTY LIST\n");
            return 0;
        }
        else{
            int value = this.tail_sentinel.previous.value;
            this.tail_sentinel.previous = this.tail_sentinel.previous.previous;
            this.tail_sentinel.previous.next = this.tail_sentinel;

            this.size--;
            return value;
        }
    }
    public int get_front(){
    /*
    Description: Function to get and return the value from the front node in the doubly linked list
    Input(s): None
    Output(s): Integer value
    */
        return this.head_sentinel.next.get_value();
    }
    public int get_back(){
    /*
    Description: Function to get and return the value from the back node in the doubly linked list
    Input(s): None
    Output(s): Integer value
    */
        return this.tail_sentinel.previous.get_value();
    }
    public class Node{
    /*
    Description: Class definition for a single node structure within the linked list.
    */
        private int value;
        public Node previous;
        public Node next;

        public Node(Node previous, Node next, int value){
        /*
        Description: Default constructor for the node structure within the linked list
        */
            this.previous = previous;
            this.next = next;
            this.value = value;
        }
        public int get_value(){
            return this.value;
        }
    }

}

